lib/event_summarizer/idv_matcher.rb

```diff
@@ -0,0 +1,436 @@
+# frozen_string_literal: true
+
+require 'active_support'
+require 'active_support/time'
+
+require 'event_summarizer/vendor_result_evaluators/aamva'
+require 'event_summarizer/vendor_result_evaluators/instant_verify'
+
+module EventSummarizer
+  class IdvMatcher
+    IDV_WELCOME_SUBMITTED_EVENT = 'IdV: doc auth welcome submitted'
+    IDV_GPO_CODE_SUBMITTED_EVENT = 'IdV: enter verify by mail code submitted'
+    IDV_FINAL_RESOLUTION_EVENT = 'IdV: final resolution'
+    IDV_VERIFY_PROOFING_RESULTS_EVENT = 'IdV: doc auth verify proofing results'
+    IPP_ENROLLMENT_STATUS_UPDATED_EVENT = 'GetUspsProofingResultsJob: Enrollment status updated'
+    PROFILE_ENCRYPTION_INVALID_EVENT = 'Profile Encryption: Invalid'
+    RATE_LIMIT_REACHED_EVENT = 'Rate Limit Reached'
+
+    EVENT_PROPERTIES = ['@message', 'properties', 'event_properties'].freeze
+
+    VENDORS = {
+      'lexisnexis:instant_verify' => {
+        id: :instant_verify,
+        name: 'Instant Verify',
+        evaluator_module: EventSummarizer::VendorResultEvaluators::InstantVerify,
+      },
+      'aamva:state_id' => {
+        id: :aamva,
+        name: 'AAMVA',
+        evaluator_module: EventSummarizer::VendorResultEvaluators::Aamva,
+      },
+    }.freeze
+
+    UNKNOWN_VENDOR = {
+      id: :unknown,
+      name: 'Unknown vendor',
+    }.freeze
+
+    IdvAttempt = Data.define(
+      :started_at,
+      :significant_events,
+    ) do
+      def initialize(started_at:, significant_events: [])
+        super(started_at:, significant_events:)
+      end
+
+      def gpo?
+        self.significant_events.any? { |e| e.type == :start_gpo }
+      end
+
+      def ipp?
+        self.significant_events.any? { |e| e.type == :start_ipp }
+      end
+
+      def successful?
+        self.significant_events.any? do |e|
+          e.type == :identity_verified
+        end
+      end
+    end.freeze
+
+    SignificantIdvEvent = Data.define(
+      :timestamp,
+      :type,
+      :description,
+    ).freeze
+
+    attr_reader :current_idv_attempt
+    attr_reader :idv_attempts
```

Comment:
When we encounter a new "welcome submitted" event, a new `IdvAttempt` is created and stored in `current_idv_attempt`. Subsequent events then update the state of that attempt. If we see _another_ "welcome submitted" event, the current_idv_attempt is put in `idv_attempts` and a new one is started.